File: examples/connext_dds/dynamic_data_skip_serialization/c++11/recorder.cxx


#include <fstream>
#include <iostream>
#include <string>
#include <rti/rti.hpp>

#include "util.hpp"

dds::core::xtypes::StructType create_type()
{
    using namespace dds::core::xtypes;

    StructType type("RecordExample", {
        Member("id", StringType(128)),
        Member("payload", SequenceType(primitive_type<int32_t>(), 1024))
    });

    return type;
}

void record(const std::string &file_name, int domain_id)
{
    using dds::core::xtypes::DynamicData;

    // To disable the deserialization on the DataReader and get direct access
    // to the serialized CDR buffer, we need to register the DataReader type
    // with the DynamicDataTypeSerializationProperty::skip_deserialization set
    // to true, and then create the DataReader with the name used to register
    // the type.
    dds::domain::DomainParticipant participant(domain_id);
    const std::string type_name = "RecordExample";
    rti::core::xtypes::DynamicDataTypeSerializationProperty property;
    property.skip_deserialization(true);
    rti::domain::register_type(participant, type_name, create_type(), property);

    dds::topic::Topic<DynamicData> topic(
            participant,
            "Example Record",
            type_name);  // specify the registered type name

    auto qos = dds::core::QosProvider::Default().datareader_qos(
            rti::core::builtin_profiles::qos_lib::generic_strict_reliable());
    dds::sub::DataReader<DynamicData> reader(topic, qos);
    if (!util::wait_for_writer(reader)) {
        return;
    }

    // File setup
    std::ofstream out_file(file_name, std::ios::binary);
    if (!out_file) {
        std::cerr << "Failed to open file for recording: " << file_name
                  << std::endl;
        return;
    }

    auto record_data = [&out_file, &reader]() {
        auto samples = reader.take();
        for (auto sample : samples) {
            if (!sample.info().valid()) {
                continue;
            }

            // Now the only way to access the data is to call get_cdr_buffer,
            // any other field accessor will fail.
            auto buffer_info = sample.data().get_cdr_buffer();
            auto buffer = buffer_info.first;
            auto buffer_length = buffer_info.second;
            std::cout << "Recording data sample (" << buffer_length << " bytes)"
                      << std::endl;
            out_file.write(
                    reinterpret_cast<const char *>(&buffer_length),
                    sizeof(buffer_length));
            out_file.write(
                    reinterpret_cast<const char *>(buffer),
                    buffer_length);
        }
    };

    // Set up a ReadCondition to trigger the record_data function when
    // data is available
    dds::core::cond::WaitSet waitset;
    dds::sub::cond::ReadCondition read_condition(
            reader,
            dds::sub::status::DataState::any(),
            record_data);
    waitset += read_condition;

    while (!application::shutdown_requested) {
        waitset.dispatch(dds::core::Duration(1));
    }
}

void replay(const std::string &file_name, int domain_id)
{
    using dds::core::xtypes::DynamicData;

    dds::domain::DomainParticipant participant(domain_id);

    // For the replay application we don't need to register the type with any
    // particular property
    dds::topic::Topic<DynamicData> topic(
            participant,
            "Example Record",
            create_type());

    auto qos = dds::core::QosProvider::Default().datawriter_qos(
            rti::core::builtin_profiles::qos_lib::generic_strict_reliable());
    dds::pub::DataWriter<DynamicData> writer(topic, qos);
    if (!util::wait_for_reader(writer)) {
        return;
    }

    std::ifstream in_file(file_name, std::ios::binary);
    if (!in_file) {
        std::cerr << "Failed to open file for replay: " << file_name
                  << std::endl;
        return;
    }

    uint32_t length;
    std::vector<char> buffer;
    DynamicData sample(create_type());
    while (!in_file.eof()) {
        // read length and data
        in_file.read(reinterpret_cast<char *>(&length), sizeof(length));

        std::cout << "Replaying data sample (" << length << " bytes)"
                  << std::endl;
        buffer.resize(length);
        in_file.read(buffer.data(), length);

        // By calling the set_cdr_buffer method we override the contents
        // of the DynamicData object with the new serialized data. After
        // setting a cdr buffer we can't use any field getters or setters.
        sample.set_cdr_buffer(buffer.data(), length);
        writer.write(sample);
    }

    in_file.close();
    writer.wait_for_acknowledgments(dds::core::Duration(10));
}

int main(int argc, char *argv[])
{
    using namespace application;

    // Parse arguments and handle control-C
    auto arguments = parse_arguments(argc, argv);
    if (arguments.parse_result == ParseReturn::exit) {
        return EXIT_SUCCESS;
    } else if (arguments.parse_result == ParseReturn::failure) {
        return EXIT_FAILURE;
    }
    setup_signal_handlers();

    // Sets Connext verbosity to help debugging
    rti::config::Logger::instance().verbosity(arguments.verbosity);

    try {
        if (arguments.application_type == ApplicationType::record) {
            record(arguments.file_name, arguments.domain_id);
        } else if (arguments.application_type == ApplicationType::replay) {
            replay(arguments.file_name, arguments.domain_id);
        } else if (arguments.application_type == ApplicationType::publish) {
            util::publish_example_data(arguments.domain_id, create_type());
        }

    } catch (const std::exception &ex) {
        std::cerr << "Exception in application: " << ex.what() << std::endl;
        return EXIT_FAILURE;
    }

    // Releases the memory used by the participant factory.  Optional at
    // application exit
    dds::domain::DomainParticipant::finalize_participant_factory();

    return EXIT_SUCCESS;
}

File: examples/connext_dds/dynamic_data_skip_serialization/py/recorder.py
# (c) 2023 Copyright, Real-Time Innovations, Inc.  All rights reserved.
#
# RTI grants Licensee a license to use, modify, compile, and create derivative
# works of the Software solely for use with RTI products.  The Software is
# provided "as is", with no warranty of any type, including any warranty for
# fitness for any purpose. RTI is under no obligation to maintain or support
# the Software.  RTI shall not be liable for any incidental or consequential
# damages arising out of the use or inability to use the software.
#

import rti.connextdds as dds
from util import wait_for_reader, wait_for_writer, publish_example_data

EXAMPLE_TYPE = dds.StructType(
    "RecordExample",
    dds.MemberSeq(
        [
            dds.Member("id", dds.StringType(128)),
            dds.Member("payload", dds.SequenceType(dds.Int32Type(), 1024)),
        ]
    ),
)


def record(file_name: str, domain_id: int):
    participant = dds.DomainParticipant(domain_id)

    # To disable the deserialization on the DataReader and get direct access
    # to the serialized CDR buffer, we need to register the DataReader type
    # with the DynamicDataTypeSerializationProperty.skip_deserialization set
    # to True, and then create the DataReader with the name used to register
    # the type.
    dd_property = dds.DynamicDataTypeSerializationProperty()
    dd_property.skip_deserialization = True
    type_name = "RecordExample"
    participant.register_type(type_name, EXAMPLE_TYPE, dd_property)

    topic = dds.DynamicData.Topic(
        participant,
        topic_name="Example Record",
        type_name=type_name,  # Specify the registered type name
        qos=participant.default_topic_qos,
    )

    qos = dds.QosProvider.default.datareader_qos_from_profile(
        dds.BuiltinProfiles.generic_strict_reliable
    )
    reader = dds.DynamicData.DataReader(topic, qos)
    wait_for_writer(reader)

    # open a file for binary writing:
    out_file = open(file_name, "wb")

    # Define the function that will take the data and record it
    def record_data(_) -> None:
        # When the skip_deserialization property is set to True the data must
        # be taken as a loaned collection, since copying the data is not allowed.
        with reader.take_loaned() as samples:
            for data, info in samples:
                if not info.valid:
                    continue

                # Now the only way to access the data is to call get_cdr_buffer,
                # any other field accessor will fail.
                buffer = data.get_cdr_buffer()
                print(f"Recording data sample ({len(buffer)} bytes)")

                # Write the length and the data
                out_file.write(len(buffer).to_bytes(4, byteorder="little"))
                out_file.write(buffer)

    # Set up a ReadCondition to trigger the record_data function when
    # data is available
    read_condition = dds.ReadCondition(reader, dds.DataState.any, record_data)
    waitset = dds.WaitSet()
    waitset += read_condition

    while True:
        try:
            waitset.dispatch(dds.Duration(1))
        except KeyboardInterrupt:
            break

    print("preparing to shut down...")


def replay(file_name: str, domain_id: int):
    participant = dds.DomainParticipant(domain_id)

    # For the replay application we don't need to register the type with any
    # particular property because DynamicData DataWriters are always prepared
    # to write serialized buffers directly
    topic = dds.DynamicData.Topic(participant, "Example Record", EXAMPLE_TYPE)

    qos = dds.QosProvider.default.datawriter_qos_from_profile(
        dds.BuiltinProfiles.generic_strict_reliable
    )
    writer = dds.DynamicData.DataWriter(topic, qos)
    wait_for_reader(writer)

    data = writer.create_data()

    with open(file_name, "rb") as in_file:
        # read the binary file and write the data to the DataWriter:
        while True:
            length_bytes = in_file.read(4)
            if length_bytes == b"":
                break

            length = int.from_bytes(length_bytes, byteorder="little")
            print(f"Replaying data sample ({length} bytes)")
            buffer = in_file.read(length)

            # By calling the set_cdr_buffer method we override the contents
            # of the DynamicData object with the new serialized data. After
            # setting a cdr buffer we can't use any field getters or setters.
            data.set_cdr_buffer(buffer)
            writer.write(data)

    writer.wait_for_acknowledgments(dds.Duration(10))


def main():
    # Parse command line for --record [output file], --replay [input file], --publish, and --domain_id options
    import argparse

    parser = argparse.ArgumentParser()
    parser.add_argument("--record", nargs="?", const="recorded_data.bin")
    parser.add_argument("--replay", nargs="?", const="recorded_data.bin")
    parser.add_argument("--publish", action="store_true")
    parser.add_argument("--domain_id", type=int, default=0)

    args = parser.parse_args()
    if args.record:
        record(args.record, args.domain_id)
    elif args.replay:
        replay(args.replay, args.domain_id)
    elif args.publish:
        publish_example_data(args.domain_id, EXAMPLE_TYPE)
    else:
        parser.print_help()


if __name__ == "__main__":
    main()
